### PR TITLE
don't compress body upon 204/304 resp status

### LINKED ISSUE #405: Avoid gzip errors for http 204 & 304 response status codes
## Is your feature request related to a problem? Please describe

The gzip middleware attempts to write to the body (the gzip header) for `204` & `304` response status codes  which aren't allowed to a body so it causes an error `http: request method or response status code does not allow body`.

`gzip.Writer.Close()` calls `gzip.Writer.Write()` https://cs.opensource.google/go/go/+/refs/tags/go1.16.7:src/compress/gzip/gzip.go;l=145-168

## Is this a bug? Please provide steps to reproduce, a failing test etc.

I've got a PR with a test, will link.

## Describe the solution

Only apply gzip middleware once we know it's applicable for the response status code

## Additional context

This is a minor problem, I'm seeing annoying errors in the logs which this change would prevent.

### PR DESCRIPTION
**Pull request by j0hnsmith**
## Which problem is this PR solving?

This PR closes #405 

## Short description of the changes

Don't apply the gzip middleware for http `204` and `304` responses which are not allowed a response body (thus preventing `http: request method or response status code does not allow body` error from occurring).

### CODE REVIEW

**Review by mantzas — DISMISSED** [review]

File: component/http/middleware.go [hunk 1]
```diff
@@ -271,29 +266,10 @@ func NewCompressionMiddleware(deflateLevel int, ignoreRoutes ...string) Middlewa
...
-				w.Header().Set(encoding.ContentEncodingHeader, deflateHeader)
-			case identityHeader, "":
-				w.Header().Set(encoding.ContentEncodingHeader, identityHeader)
-				fallthrough
-			// `*`, `identity` and others must fall through here to be served without compression
-			default:
-				next.ServeHTTP(w, r)
-				return
-			}
+			dw := &dynamicCompressionResponseWriter{w, selectedEncoding, nil, 0, deflateLevel}
```
Comment (tpaschalis) [inline component/http/middleware.go:269]: If we do

**Review by tpaschalis — DISMISSED** [review]

**Comment by j0hnsmith** [comment]
Rebased against master in the hope that the failing test will be solved.

**Comment by j0hnsmith** [comment]
Feel free to merge once this approved as I may not notice

**Review by mantzas — APPROVED** [review]

**Review by tpaschalis — APPROVED** [review]

**Comment by j0hnsmith** [comment]
@tpaschalis do you know why the build is failing, I'm pretty sure it's not related to this PR. I can rebase if that will make the build pass

**Comment by tpaschalis** [comment]
Not sure why CI fails tbh. I re-ran it to see if it's just something flaky or else I'll try to find some time and take a better look.

**Pull request merged**